File: packages/python-sdk/src/neon3_sdk/input.py
"""Input contracts. Pointer injection is available; keyboard requires a negotiated runtime capability."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from .client import NeonClient
from .errors import RemoteError
from .render import PointerEvent


@dataclass(frozen=True)
class KeyEvent:
    key: str
    action: str
    modifiers: tuple[str, ...] = ()
    repeat: bool = False

    def to_wire(self) -> dict[str, Any]:
        if self.action not in {"down", "up"}:
            raise ValueError("keyboard action must be down or up")
        return {"key": self.key, "action": self.action, "modifiers": list(self.modifiers), "repeat": self.repeat}


class InputClient:
    def __init__(self, client: NeonClient, target: str = "wgpu-runtime") -> None:
        self.client = client
        self.target = target

    def pointer(self, event: PointerEvent) -> Any:
        return self.client.call(self.target, "ui.host.pointer_event", {"event": event.to_wire()}).result

    def keyboard(self, event: KeyEvent) -> Any:
        """Send a typed keyboard event when the runtime advertises that contract."""
        description = self.client.describe(self.target)
        if "wgpu.ui.keyboard.v1" not in description.capabilities:
            raise RemoteError("keyboard-capability", "rejected", {"code": "keyboard_capability_unavailable", "message": "The connected Neon3 runtime does not advertise wgpu.ui.keyboard.v1."})
        return self.client.call(self.target, "ui.host.keyboard_event", {"event": event.to_wire()}).result

    def debug_snapshot(self) -> Any:
        return self.client.call(self.target, "debug.window.input.snapshot").result

File: packages/node-sdk/src/input.ts
import { NeonClient } from "./client.js";
import { RemoteError } from "./errors.js";
import { PointerEvent, PointerEventWire } from "./render.js";

export interface KeyEvent {
  key: string;
  action: "down" | "up";
  modifiers?: string[];
  repeat?: boolean;
}

export function keyEventToWire(event: KeyEvent): Record<string, unknown> {
  if (event.action !== "down" && event.action !== "up") throw new Error("keyboard action must be down or up");
  return { key: event.key, action: event.action, modifiers: event.modifiers ?? [], repeat: event.repeat ?? false };
}

export class InputClient {
  constructor(readonly client: NeonClient, readonly target = "wgpu-runtime") {}

  async pointer(event: PointerEvent | PointerEventWire): Promise<unknown> {
    const wire = event instanceof PointerEvent ? event.toWire() : event;
    return (await this.client.call(this.target, "ui.host.pointer_event", { event: wire })).result;
  }

  async keyboard(event: KeyEvent): Promise<unknown> {
    const description = await this.client.describe(this.target);
    if (!description.capabilities.includes("wgpu.ui.keyboard.v1")) {
      throw new RemoteError("keyboard-capability", "rejected", { code: "keyboard_capability_unavailable", message: "runtime does not advertise wgpu.ui.keyboard.v1" });
    }
    return (await this.client.call(this.target, "ui.host.keyboard_event", { event: keyEventToWire(event) })).result;
  }

  async debugSnapshot(): Promise<unknown> {
    return (await this.client.call(this.target, "debug.window.input.snapshot")).result;
  }
}
